File: Chapter02/Activity2_03/Activity2_03.scala
package Activity2_03

import Utilities02.HelperScala.{extractRawRecords, parseRawWarc}
import Activity2_02.Activity2_02.{extractPlainText, detectLanguage}
import Utilities02.WarcRecord
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession

object Activity2_03 {

  // ~/spark-2.4.6-bin-hadoop2.7/bin/spark-submit --master local[1] --class Activity2_03.Activity2_03 --driver-class-path /Users/a/.m2/repository/com/google/guava/guava/28.2-jre/guava-28.2-jre.jar:/Users/a/.m2/repository/org/apache/commons/commons-compress/1.20/commons-compress-1.20.jar ~/IdeaProjects/The-Spark-Workshop/target/packt-uber-jar.jar ~/IdeaProjects/The-Spark-Workshop/resources/webcorpus/warc.sample
  def main(args: Array[String]): Unit = {
    implicit val session = SparkSession.builder
      .appName("Crawl Tagger Estimation")
      .getOrCreate()

    val input = args(0)
    val warcRecords: RDD[WarcRecord] = extractRawRecords(input)
      .flatMap(parseRawWarc)
      .filter(_.warcType == "response")

    val taggedTexts = warcRecords.flatMap(record => {
      val plainText = extractPlainText(record)
      if (plainText.isDefined)
        Some(record.targetURI, plainText.get)
      else
        None
    }).map { case (uri, text) =>
      val (language, confidence) = detectLanguage(text)
      (uri, language, confidence)
    }

    session.time(taggedTexts.count())

  }
}

File: Chapter02/Activity2_03/Activity2_03.py
from sys import argv
from pyspark.sql import SparkSession
from Chapter02.Activity2_02.Activity2_02 import extract_plaintext, detect_language
from Chapter02.utilities02_py.helper_python import extract_raw_records, parse_raw_warc

#   ~/spark-2.4.6-bin-hadoop2.7/bin/spark-submit --master local[1] ~/IdeaProjects/The-Spark-Workshop/Chapter02/Activity2_03/Activity2_03.py ~/IdeaProjects/The-Spark-Workshop/resources/webcorpus/warc.sample
if __name__ == "__main__":
    session: SparkSession = SparkSession.builder \
        .appName('Crawl Tagger Estimation') \
        .getOrCreate()
    input_file = argv[1]
    warc_records = extract_raw_records(input_file, session) \
        .flatMap(lambda record: parse_raw_warc(record)) \
        .filter(lambda record: record.warc_type == 'response')

    plaintexts_rdd = warc_records.flatMap(extract_plaintext)
    tagged_texts_rdd = plaintexts_rdd.map(lambda record: (record[0], detect_language(record[1])))

    from timeit import default_timer
    start_time = default_timer()
    size = tagged_texts_rdd.count()
    end_time = default_timer()
    print('Time taken: ' + str(end_time - start_time))
